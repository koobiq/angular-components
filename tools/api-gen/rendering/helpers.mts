import { ClassEntryRenderable, DocEntryRenderable } from './entities/renderables.mjs';
import { isClassEntry, isDeprecatedEntry } from './entities/categorization.mjs';
import { EntryType } from './entities.mjs';

export const isClass = (renderable: DocEntryRenderable) => {
    return isClassEntry(renderable) && renderable.entryType === EntryType.UndecoratedClass
        && !(renderable as ClassEntryRenderable).isService;
}

export const isDirective = (renderable: DocEntryRenderable): boolean => {
    return renderable.entryType === EntryType.Component ||
        renderable.entryType === EntryType.Directive;
}

export function findBestPrimaryExport(docs: DocEntryRenderable[]): DocEntryRenderable | null {
    // Usually the first doc that is not deprecated is used, but in case there are
    // only deprecated doc, the last deprecated doc is used. We don't want to always
    // skip deprecated docs as they could be still needed for documentation of a
    // deprecated entry-point.
    for (const doc of docs) {
        if (!isDeprecatedEntry(doc)) { return doc; }
    }

    return null;
}

export function entryPointGrouper(renderables: DocEntryRenderable[], moduleName: string, packageName: string) {
    const packageDisplayName = moduleName === 'cdk' ? 'CDK' : 'Koobiq';
    const moduleImportPath = `@koobiq/${moduleName}/${packageName}`;
    const exportedNgModules = renderables.filter((renderable) => renderable.entryType === EntryType.NgModule);
    const ngModuleExport = findBestPrimaryExport(exportedNgModules);

    return {
        primaryExportName: ngModuleExport?.name || '',
        displayName: packageName,
        moduleImportPath,
        packageDisplayName,
        exportedNgModules,
        directives: renderables.filter(isDirective),
        classes: renderables.filter(isClass),
        interfaces: renderables.filter((renderable) => renderable.entryType === EntryType.Interface),
        typeAliases: renderables.filter((renderable) => renderable.entryType === EntryType.TypeAlias),
        constants: renderables.filter((renderable) => renderable.entryType === EntryType.Constant),
        functions: renderables.filter((renderable) => renderable.entryType === EntryType.Function),
        services: renderables.filter((renderable) => isClassEntry(renderable) && renderable.isService),
    }
}
