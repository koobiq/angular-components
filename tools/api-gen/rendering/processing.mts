import { DocEntryRenderable } from './entities/renderables.mjs';
import {
    isClassEntry,
    isConstantEntry,
    isEnumEntry,
    isFunctionEntry,
    isInterfaceEntry,
    isTypeAliasEntry
} from './entities/categorization.mjs';

import { getClassRenderable } from './transforms/class-transforms.mjs';
import { getConstantRenderable } from './transforms/constant-transforms.mjs';
import { getEnumRenderable } from './transforms/enum-transforms.mjs';
import { getInterfaceRenderable } from './transforms/interface-transforms.mjs';
import { getFunctionRenderable } from './transforms/function-transforms.mjs';
import { getTypeAliasRenderable } from './transforms/type-alias-transforms.mjs';

import {
    addHtmlAdditionalLinks,
    addHtmlDescription, addHtmlJsDocTagComments,
    addHtmlUsageNotes,
    setEntryFlags
} from './transforms/jsdoc-transforms.mjs';
import { addModuleName } from './transforms/module-name.mjs';
import { DocEntry } from './entities.mjs';

export function getRenderable(entry: DocEntry, moduleName: string): DocEntryRenderable {
    if (isClassEntry(entry)) {
        return getClassRenderable(entry, moduleName);
    }
    if (isConstantEntry(entry)) {
        return getConstantRenderable(entry, moduleName);
    }
    if (isEnumEntry(entry)) {
        return getEnumRenderable(entry, moduleName);
    }
    if (isInterfaceEntry(entry)) {
        return getInterfaceRenderable(entry, moduleName);
    }
    if (isFunctionEntry(entry)) {
        return getFunctionRenderable(entry, moduleName);
    }
    if (isTypeAliasEntry(entry)) {
        return getTypeAliasRenderable(entry, moduleName);
    }

    /*if (isInitializerApiFunctionEntry(entry)) {
        return getInitializerApiFunctionRenderable(entry, moduleName);
    }
    if (isCliEntry(entry)) {
        return getCliRenderable(entry);
    }*/

    // Fallback to an uncategorised renderable.
    return setEntryFlags(
        addHtmlAdditionalLinks(
            addHtmlDescription(
                addHtmlUsageNotes(addHtmlJsDocTagComments(addModuleName(entry, moduleName))),
            ),
        ),
    );
}
