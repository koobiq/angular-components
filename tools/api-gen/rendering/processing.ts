import { DocEntryRenderable } from './entities/renderables.ts';
import {
    isClassEntry,
    isConstantEntry,
    isEnumEntry,
    isFunctionEntry,
    isInterfaceEntry,
    isTypeAliasEntry
} from './entities/categorization.ts';

import { getClassRenderable } from './transforms/class-transforms.ts';
import { getConstantRenderable } from './transforms/constant-transforms.ts';
import { getEnumRenderable } from './transforms/enum-transforms.ts';
import { getInterfaceRenderable } from './transforms/interface-transforms.ts';
import { getFunctionRenderable } from './transforms/function-transforms.ts';
import { getTypeAliasRenderable } from './transforms/type-alias-transforms.ts';

import {
    addHtmlAdditionalLinks,
    addHtmlDescription, addHtmlJsDocTagComments,
    addHtmlUsageNotes,
    setEntryFlags
} from './transforms/jsdoc-transforms.ts';
import { addModuleName } from './transforms/module-name.ts';
import { DocEntry } from './entities.ts';

export function getRenderable(entry: DocEntry, moduleName: string): DocEntryRenderable {
    if (isClassEntry(entry)) {
        return getClassRenderable(entry, moduleName);
    }
    if (isConstantEntry(entry)) {
        return getConstantRenderable(entry, moduleName);
    }
    if (isEnumEntry(entry)) {
        return getEnumRenderable(entry, moduleName);
    }
    if (isInterfaceEntry(entry)) {
        return getInterfaceRenderable(entry, moduleName);
    }
    if (isFunctionEntry(entry)) {
        return getFunctionRenderable(entry, moduleName);
    }
    if (isTypeAliasEntry(entry)) {
        return getTypeAliasRenderable(entry, moduleName);
    }

    /*if (isInitializerApiFunctionEntry(entry)) {
        return getInitializerApiFunctionRenderable(entry, moduleName);
    }
    if (isCliEntry(entry)) {
        return getCliRenderable(entry);
    }*/

    // Fallback to an uncategorized renderable.
    return setEntryFlags(
        addHtmlAdditionalLinks(
            addHtmlDescription(
                addHtmlUsageNotes(addHtmlJsDocTagComments(addModuleName(entry, moduleName))),
            ),
        ),
    );
}
