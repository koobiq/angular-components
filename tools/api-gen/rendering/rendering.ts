import nunjucks from 'nunjucks';
import { join } from 'path';
import { HighlightNunjucksExtension } from './nunjucks-tags/highlight';

const TEMPLATE_DIR = join(process.cwd(), 'tools', 'api-gen', 'rendering', 'templates');
const env = nunjucks.configure(TEMPLATE_DIR, {
    autoescape: false,
    tags: {
        blockStart: '{%',
        blockEnd: '%}',
        variableStart: '{$',
        variableEnd: '$}'
    }
});
env.addExtension('HighlightJsExtension', new HighlightNunjucksExtension());

export function renderEntry(entryPointContext: any) {
    try {
        return env.getTemplate('entry-point.template.html').render({ doc: entryPointContext });
    } catch (e) {
        console.warn(e);
        return '';
    }
}
