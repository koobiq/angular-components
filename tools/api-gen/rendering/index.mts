import { writeFileSync } from 'fs';

import { configureMarkedGlobally } from './marked/configuration.mjs';
import { getRenderable } from './processing.mjs';
import { renderEntry } from './rendering.mjs';
import { createDirIfNotExists } from '../utils.mjs';
import { EntryCollection, ManifestEntry, PackageApiInfo } from '../types/index.mjs';
import { entryPointGrouper } from './helpers.mjs';

export function generateApiToHtml(entryCollections: EntryCollection<ManifestEntry>[]) {
    configureMarkedGlobally();

    for (const collection of entryCollections) {
        collection.packagesApiInfo.forEach((packageApiInfo: PackageApiInfo) => {
            const extractedEntries = packageApiInfo.entries;
            const renderableEntries = extractedEntries.map((entry) =>
                getRenderable(entry, collection.moduleName)
            )

            const entryPointContext = entryPointGrouper(
                renderableEntries, collection.moduleName, packageApiInfo.packageName
            );

            const htmlOutput = renderEntry(entryPointContext);

            createDirIfNotExists('dist/docs-content/api-docs').then(() => {
                writeFileSync(
                    `dist/docs-content/api-docs/${ collection.moduleName }-${ packageApiInfo.packageName }.html`,
                    htmlOutput,
                    { encoding: 'utf8' });
            });
        })
    }
}
