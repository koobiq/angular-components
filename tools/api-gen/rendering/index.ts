import { writeFileSync } from 'fs';

import { configureMarkedGlobally } from './marked/configuration.ts';
import { getRenderable } from './processing.ts';
import { renderEntry } from './rendering.ts';
import { createDirIfNotExists } from '../utils.ts';
import { EntryCollection, ManifestEntry, PackageApiInfo } from '../types/types.ts';
import { entryPointGrouper } from './helpers.ts';

export function generateApiToHtml(entryCollections: EntryCollection<ManifestEntry>[]) {
    configureMarkedGlobally();

    for (const collection of entryCollections) {
        collection.packagesApiInfo.forEach((packageApiInfo: PackageApiInfo) => {
            const extractedEntries = packageApiInfo.entries;
            const renderableEntries = extractedEntries.map((entry) =>
                getRenderable(entry, collection.moduleName)
            )

            const entryPointContext = entryPointGrouper(
                renderableEntries, collection.moduleName, packageApiInfo.packageName
            );

            const htmlOutput = renderEntry(entryPointContext);

            createDirIfNotExists('dist/docs-content/api-docs').then(() => {
                writeFileSync(
                    `dist/docs-content/api-docs/${ collection.moduleName }-${ packageApiInfo.packageName }.html`,
                    htmlOutput,
                    { encoding: 'utf8' });
            });
        })
    }
}
