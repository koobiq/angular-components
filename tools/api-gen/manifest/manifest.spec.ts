import { DocEntry, EntryType } from '../rendering/entities';
import { generateManifest } from './';

describe('api manifest generation', () => {
    it('should generate a manifest from multiple collections', () => {
        expect(
            generateManifest([
                {
                    moduleName: 'cdk',
                    packagesApiInfo: [
                        {
                            packageName: 'button',
                            entries: [
                                entry({ name: 'KbqButton', entryType: EntryType.Component }),
                                entry({ name: 'TOOLTIP_TIMEOUT', entryType: EntryType.Constant })
                            ]
                        }
                    ]
                }
            ])
        ).toMatchSnapshot();
    });

    it('should should filter private entries', () => {
        expect(
            generateManifest([
                {
                    moduleName: 'components',
                    packagesApiInfo: [
                        {
                            packageName: 'button',
                            entries: [
                                entry({ name: 'KbqButton', entryType: EntryType.Component }),
                                entry({
                                    name: 'PRIVATE_CONST',
                                    entryType: EntryType.Constant,
                                    jsdocTags: [{ name: 'docs-private', comment: 'test' }]
                                })

                            ]
                        }
                    ]
                }
            ])
        ).toMatchSnapshot();
    });
});

/** Creates a fake DocsEntry with the given object's fields patches onto the result. */
function entry(patch: Partial<DocEntry>): DocEntry {
    return {
        name: '',
        description: '',
        entryType: EntryType.Constant,
        jsdocTags: [],
        rawComment: '',
        ...patch
    };
}
