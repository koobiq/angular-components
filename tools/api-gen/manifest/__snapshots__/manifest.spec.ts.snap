// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api manifest generation should generate a manifest from multiple collections 1`] = `
[
  {
    "moduleName": "cdk",
    "packagesApiInfo": [
      {
        "entries": [
          {
            "description": "",
            "entryType": "component",
            "jsdocTags": [],
            "name": "KbqButton",
            "publicUrl": "cdk/button/api#KbqButton",
            "rawComment": "",
          },
          {
            "description": "",
            "entryType": "constant",
            "jsdocTags": [],
            "name": "TOOLTIP_TIMEOUT",
            "publicUrl": "cdk/button/api#TOOLTIP_TIMEOUT",
            "rawComment": "",
          },
        ],
        "packageName": "button",
      },
    ],
  },
]
`;

exports[`api manifest generation should should filter private entries 1`] = `
[
  {
    "moduleName": "components",
    "packagesApiInfo": [
      {
        "entries": [
          {
            "description": "",
            "entryType": "component",
            "jsdocTags": [],
            "name": "KbqButton",
            "publicUrl": "components/button/api#KbqButton",
            "rawComment": "",
          },
        ],
        "packageName": "button",
      },
    ],
  },
]
`;
